### PR TITLE
Add time stamps for wav2vec2 with lm

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

This PR adds word offsets analogue to https://github.com/huggingface/transformers/pull/15687 . `pyctcedecode` only returns time stamps of words not characters so only those can be returned for now.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/models/wav2vec2_with_lm/processing_wav2vec2_with_lm.py [hunk 1]
```diff
@@ -390,9 +473,18 @@ def decode(
             hotword_weight=hotword_weight,
         )
 
+        word_offsets = None
+        if output_word_offsets:
+            word_offsets = [
+                {"word": t[0], "start_offset": t[1][0], "end_offset": t[1][1]} for t in decoded_beams[0][2]
```
Comment (Narsil) [inline src/transformers/models/wav2vec2_with_lm/processing_wav2vec2_with_lm.py:479]:
```suggestion
                {"word": word, "start_offset": start_offset, "end_offset": end_offset} for word, (start_offset, end_offset) in decoded_beams[0][2]
```
Nit for readability

**Review by Narsil — APPROVED** [review]
Very nice ! 

LGTM

**Comment by patrickvonplaten** [comment]
@sgugger - if you could take a quick look here this would be great :-)

File: src/transformers/models/wav2vec2/tokenization_wav2vec2.py [hunk 2]
```diff
@@ -106,18 +108,18 @@ class Wav2Vec2CTCTokenizerOutput(ModelOutput):
     Args:
         text (list of `str` or `str`):
             Decoded logits in text from. Usually the speech transcription.
-        char_offsets (`Dict[str, Union[int, str]]` or `Dict[str, Union[int, str]]`):
+        char_offsets (list of `List[Dict[str, Union[int, str]]]` or `List[Dict[str, Union[int, str]]]`):
             Offsets of the decoded characters. In combination with sampling rate and model downsampling rate char
             offsets can be used to compute time stamps for each charater. Total logit score of the beam associated with
             produced text.
-        word_offsets (`Dict[str, Union[int, str]]` or `Dict[str, Union[int, str]]`):
+        word_offsets (list of `List[Dict[str, Union[int, str]]]` or `List[Dict[str, Union[int, str]]]`):
```
Comment (sgugger) [inline src/transformers/models/wav2vec2/tokenization_wav2vec2.py:115]: The type alias corresponds to `Union[float,str]`

File: src/transformers/models/wav2vec2_phoneme/tokenization_wav2vec2_phoneme.py [hunk 3]
```diff
@@ -74,14 +77,14 @@ class Wav2Vec2PhonemeCTCTokenizerOutput(ModelOutput):
     Args:
         text (list of `str` or `str`):
             Decoded logits in text from. Usually the speech transcription.
-        char_offsets (`Dict[str, Union[int, str]]` or `Dict[str, Union[int, str]]`):
+        char_offsets (list of `List[Dict[str, Union[int, str]]]` or `List[Dict[str, Union[int, str]]]`):
```
Comment (sgugger) [inline src/transformers/models/wav2vec2_phoneme/tokenization_wav2vec2_phoneme.py:80]: Same here.

File: src/transformers/models/wav2vec2_with_lm/processing_wav2vec2_with_lm.py [hunk 4]
```diff
@@ -310,13 +330,21 @@ def batch_decode(
...
+            word_offsets.append([{"word": t[0], "start_offset": t[1][0], "end_offset": t[1][1]} for t in d[0][1]])
+
+        word_offsets = word_offsets if output_word_offsets else None
+
+        return Wav2Vec2DecoderWithLMOutput(
+            text=batch_texts,
+            logit_score=logit_scores,
+            lm_score=lm_scores,
+            word_offsets=word_offsets,
+        )
```
Comment (sgugger) [inline src/transformers/models/wav2vec2_with_lm/processing_wav2vec2_with_lm.py:347]:
```suggestion
        return Wav2Vec2DecoderWithLMOutput(
            text=batch_texts, logit_score=logit_scores, lm_score=lm_scores, word_offsets=word_offsets
        )
```

File: tests/wav2vec2_with_lm/test_processor_wav2vec2_with_lm.py [hunk 5]
```diff
@@ -350,3 +356,102 @@ def test_processor_from_auto_processor(self):
...
+        import torch
+
+        ds = load_dataset("common_voice", "en", split="train", streaming=True)
+        ds = ds.cast_column("audio", datasets.Audio(sampling_rate=16_000))
+        ds_iter = iter(ds)
+        sample = next(ds_iter)
+
+        processor = AutoProcessor.from_pretrained("patrickvonplaten/wav2vec2-base-100h-with-lm")
+        model = Wav2Vec2ForCTC.from_pretrained("patrickvonplaten/wav2vec2-base-100h-with-lm")
+        # feature_extractor = AutoFeatureExtractor.from_pretrained("facebook/wav2vec2-base-960h")
```
Comment (sgugger) [inline tests/wav2vec2_with_lm/test_processor_wav2vec2_with_lm.py:413]: To clean up?

**Review by sgugger — APPROVED** [review]
Thanks for adding this!

File: src/transformers/models/wav2vec2_phoneme/tokenization_wav2vec2_phoneme.py [hunk 6]
```diff
@@ -66,6 +66,9 @@
 PRETRAINED_POSITIONAL_EMBEDDINGS_SIZES = {"facebook/wav2vec2-lv-60-espeak-cv-ft": sys.maxsize}
 
 
+ListOfDict = List[Dict[str, Union[float, str]]]
```
Comment (patrickvonplaten) [inline src/transformers/models/wav2vec2_phoneme/tokenization_wav2vec2_phoneme.py:69]:
```suggestion
ListOfDict = List[Dict[str, Union[int, str]]]
```

File: src/transformers/models/wav2vec2/tokenization_wav2vec2.py [hunk 7]
```diff
@@ -97,6 +97,8 @@
                 Whether or not to print more information and warnings.
 """
 
+ListOfDict = List[Dict[str, Union[float, str]]]
```
Comment (patrickvonplaten) [inline src/transformers/models/wav2vec2/tokenization_wav2vec2.py:100]:
```suggestion
ListOfDict = List[Dict[str, Union[int, str]]]
```

File: src/transformers/models/wav2vec2_with_lm/processing_wav2vec2_with_lm.py [hunk 8]
```diff
@@ -34,23 +34,30 @@
     from ...tokenization_utils import PreTrainedTokenizerBase
 
 
+ListOfDict = List[Dict[str, Union[float, str]]]
```
Comment (patrickvonplaten) [inline src/transformers/models/wav2vec2_with_lm/processing_wav2vec2_with_lm.py:37]:
```suggestion
ListOfDict = List[Dict[str, Union[int, str]]]
```

File: tests/wav2vec2_with_lm/test_processor_wav2vec2_with_lm.py (hunk 5 above)
Comment (patrickvonplaten) [inline tests/wav2vec2_with_lm/test_processor_wav2vec2_with_lm.py:413]:
```suggestion
```

File: src/transformers/models/wav2vec2/tokenization_wav2vec2.py (hunk 2 above)
Comment (patrickvonplaten) [reply src/transformers/models/wav2vec2/tokenization_wav2vec2.py:115]: should have been `int` - thanks for spotting this

**Pull request merged**